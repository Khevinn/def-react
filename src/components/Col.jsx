import { useMemo } from 'react'
import useClassNames from "../hooks/useClassNames"
import useStyleProp from "../hooks/useStyleProp"

const Col = ( {
  className: classNameProp,
  size, sm, md, lg, xl,
  ...props
} ) => {
  const bootstrapClassSize = useMemo( () => {
    return Object.fromEntries(
      Object.entries( { size, sm, md, lg, xl } )
        .filter( ( [ key, value ] ) => !!value )
        .map( ( [ key, value ] ) => 
          key === 'size'
            ? [ `col-${value}`, true ] 
            : [ `col-${key}-${value}`, true ]
        )
    )
  }, [ size, sm, md, lg, xl ] )
  const className = useClassNames( 'col', bootstrapClassSize, classNameProp )
  return <div className={className} {...useStyleProp( props )} />
}

export default Col
