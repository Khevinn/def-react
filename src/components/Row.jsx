import useClassNames from "../hooks/useClassNames"
import useStyleProp from "../hooks/useStyleProp"

const Row = ( { className: classNameProp, ...props } ) => {
  const className = useClassNames( 'row', classNameProp )
  return <div className={className} {...useStyleProp( props )} />
}

export default Row
