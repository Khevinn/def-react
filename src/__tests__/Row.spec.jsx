import { create, act } from 'react-test-renderer'
import Row from '../components/Row'
import Col from '../components/Col'

describe( 'row match props', () => {
  it( 'compare snapshot', () => {
    let reactTestInstace
    act( () => {
      reactTestInstace = create(
        <Row margin={10} padding={20}>
          <Col margin={15}>
            <h1>Hello World</h1>
          </Col>
          <Col margin={20}>
            <h1>I'am column</h1>
          </Col>
        </Row>
      )
    } )

    const omitedProps = ( () => {
      const { children, ...rest } = reactTestInstace.root.props
      return rest
    } )()

    expect( omitedProps ).toEqual( { margin: 10, padding: 20 } )

    expect( reactTestInstace.root.type ).toBe( Row )

    expect( reactTestInstace.root.parent ).toBeNull()

    reactTestInstace.root.findAllByProps( { className: 'col' } )

    reactTestInstace.root.findAllByType( 'div' )

    reactTestInstace.root.findAllByType( Col ).forEach( ( child ) => {
      expect( child.type ).toBe( Col )
      expect( child.parent.type ).toBe( 'div' )
      expect( child.parent.parent ).toBe( reactTestInstace.root )
      expect( child.parent.parent.type ).toBe( Row )
      expect( child.parent.parent.parent ).toBeNull()
    } )

    reactTestInstace.root.findByProps( { margin: 15 } )

    const findByPropMargin20 = reactTestInstace.root.findByProps( { margin: 20 } )

    expect( findByPropMargin20 ).toBeDefined()

    expect( findByPropMargin20.type ).toBe( Col )

    expect( reactTestInstace.toJSON() ).toMatchSnapshot()

    expect( reactTestInstace.toTree() ).toMatchSnapshot( 'tree' )

    act( () => {
      reactTestInstace.update(
        <Row padding={40}>
          <Col margin={15}>
            <h1>Hello World</h1>
          </Col>
          <Col>
            <h1>I'am column</h1>
          </Col>
          <Col padding={15}>
            <h1>I'am last column</h1>
          </Col>
        </Row>
      )
    } )

    expect( reactTestInstace.root.props.padding ).toEqual( 40 )

    expect( reactTestInstace.root.children[0].children.length ).toBe( 3 )

    expect( reactTestInstace.toJSON() ).toMatchSnapshot()

    expect( reactTestInstace.toTree() ).toMatchSnapshot( 'tree' )
  } )
} )
