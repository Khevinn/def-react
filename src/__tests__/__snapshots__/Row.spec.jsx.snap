// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`row match props compare snapshot 1`] = `
<div
  className="row"
  style={
    Object {
      "margin": 10,
      "padding": 20,
    }
  }
>
  <div
    className="col"
    style={
      Object {
        "margin": 15,
        "padding": undefined,
      }
    }
  >
    <h1>
      Hello World
    </h1>
  </div>
  <div
    className="col"
    style={
      Object {
        "margin": 20,
        "padding": undefined,
      }
    }
  >
    <h1>
      I'am column
    </h1>
  </div>
</div>
`;

exports[`row match props compare snapshot 2`] = `
<div
  className="row"
  style={
    Object {
      "margin": undefined,
      "padding": 40,
    }
  }
>
  <div
    className="col"
    style={
      Object {
        "margin": 15,
        "padding": undefined,
      }
    }
  >
    <h1>
      Hello World
    </h1>
  </div>
  <div
    className="col"
    style={
      Object {
        "margin": undefined,
        "padding": undefined,
      }
    }
  >
    <h1>
      I'am column
    </h1>
  </div>
  <div
    className="col"
    style={
      Object {
        "margin": undefined,
        "padding": 15,
      }
    }
  >
    <h1>
      I'am last column
    </h1>
  </div>
</div>
`;

exports[`row match props compare snapshot: tree 1`] = `
Object {
  "instance": null,
  "nodeType": "component",
  "props": Object {
    "children": Array [
      <Col
        margin={15}
      >
        <h1>
          Hello World
        </h1>
      </Col>,
      <Col
        margin={20}
      >
        <h1>
          I'am column
        </h1>
      </Col>,
    ],
    "margin": 10,
    "padding": 20,
  },
  "rendered": Object {
    "instance": null,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Col
          margin={15}
        >
          <h1>
            Hello World
          </h1>
        </Col>,
        <Col
          margin={20}
        >
          <h1>
            I'am column
          </h1>
        </Col>,
      ],
      "className": "row",
      "style": Object {
        "margin": 10,
        "padding": 20,
      },
    },
    "rendered": Array [
      Object {
        "instance": null,
        "nodeType": "component",
        "props": Object {
          "children": <h1>
            Hello World
          </h1>,
          "margin": 15,
        },
        "rendered": Object {
          "instance": null,
          "nodeType": "host",
          "props": Object {
            "children": <h1>
              Hello World
            </h1>,
            "className": "col",
            "style": Object {
              "margin": 15,
              "padding": undefined,
            },
          },
          "rendered": Array [
            Object {
              "instance": null,
              "nodeType": "host",
              "props": Object {
                "children": "Hello World",
              },
              "rendered": Array [
                "Hello World",
              ],
              "type": "h1",
            },
          ],
          "type": "div",
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "nodeType": "component",
        "props": Object {
          "children": <h1>
            I'am column
          </h1>,
          "margin": 20,
        },
        "rendered": Object {
          "instance": null,
          "nodeType": "host",
          "props": Object {
            "children": <h1>
              I'am column
            </h1>,
            "className": "col",
            "style": Object {
              "margin": 20,
              "padding": undefined,
            },
          },
          "rendered": Array [
            Object {
              "instance": null,
              "nodeType": "host",
              "props": Object {
                "children": "I'am column",
              },
              "rendered": Array [
                "I'am column",
              ],
              "type": "h1",
            },
          ],
          "type": "div",
        },
        "type": [Function],
      },
    ],
    "type": "div",
  },
  "type": [Function],
}
`;

exports[`row match props compare snapshot: tree 2`] = `
Object {
  "instance": null,
  "nodeType": "component",
  "props": Object {
    "children": Array [
      <Col
        margin={15}
      >
        <h1>
          Hello World
        </h1>
      </Col>,
      <Col>
        <h1>
          I'am column
        </h1>
      </Col>,
      <Col
        padding={15}
      >
        <h1>
          I'am last column
        </h1>
      </Col>,
    ],
    "padding": 40,
  },
  "rendered": Object {
    "instance": null,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Col
          margin={15}
        >
          <h1>
            Hello World
          </h1>
        </Col>,
        <Col>
          <h1>
            I'am column
          </h1>
        </Col>,
        <Col
          padding={15}
        >
          <h1>
            I'am last column
          </h1>
        </Col>,
      ],
      "className": "row",
      "style": Object {
        "margin": undefined,
        "padding": 40,
      },
    },
    "rendered": Array [
      Object {
        "instance": null,
        "nodeType": "component",
        "props": Object {
          "children": <h1>
            Hello World
          </h1>,
          "margin": 15,
        },
        "rendered": Object {
          "instance": null,
          "nodeType": "host",
          "props": Object {
            "children": <h1>
              Hello World
            </h1>,
            "className": "col",
            "style": Object {
              "margin": 15,
              "padding": undefined,
            },
          },
          "rendered": Array [
            Object {
              "instance": null,
              "nodeType": "host",
              "props": Object {
                "children": "Hello World",
              },
              "rendered": Array [
                "Hello World",
              ],
              "type": "h1",
            },
          ],
          "type": "div",
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "nodeType": "component",
        "props": Object {
          "children": <h1>
            I'am column
          </h1>,
        },
        "rendered": Object {
          "instance": null,
          "nodeType": "host",
          "props": Object {
            "children": <h1>
              I'am column
            </h1>,
            "className": "col",
            "style": Object {
              "margin": undefined,
              "padding": undefined,
            },
          },
          "rendered": Array [
            Object {
              "instance": null,
              "nodeType": "host",
              "props": Object {
                "children": "I'am column",
              },
              "rendered": Array [
                "I'am column",
              ],
              "type": "h1",
            },
          ],
          "type": "div",
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "nodeType": "component",
        "props": Object {
          "children": <h1>
            I'am last column
          </h1>,
          "padding": 15,
        },
        "rendered": Object {
          "instance": null,
          "nodeType": "host",
          "props": Object {
            "children": <h1>
              I'am last column
            </h1>,
            "className": "col",
            "style": Object {
              "margin": undefined,
              "padding": 15,
            },
          },
          "rendered": Array [
            Object {
              "instance": null,
              "nodeType": "host",
              "props": Object {
                "children": "I'am last column",
              },
              "rendered": Array [
                "I'am last column",
              ],
              "type": "h1",
            },
          ],
          "type": "div",
        },
        "type": [Function],
      },
    ],
    "type": "div",
  },
  "type": [Function],
}
`;
